### DISCUSSION TITLE
Java external functions with JIT

### DISCUSSION THREAD

**Discussion by agobL** [jax-ml/jax#12312] [Q&A] [answered]
Hi,

I succeeded to run compiled Java functions in external functions in Python without using JIT.
But when I activate JIT, it does not work (compiled Java functions does not appreciate ShapedArray elements).
The error I got is "ConcretizationTypeError" (it needs the concrete value of an array) .
Is there a trick to make that running possible with JIT (in order to increase the rapidity of the code) ?

Thanks in advance.

**Comment by jakevdp** [comment, marked as answer]
The best approach would probably be to call your functions via the new [`jax.pure_callback`](https://jax.readthedocs.io/en/latest/_autosummary/jax.pure_callback.html?highlight=pure_callback) mechanism (available in jax version 0.3.17 or newer).

Documentation for this is still relatively incomplete (we're working on it!) but the example here may be helpful: https://github.com/google/jax/discussions/12180

**Reply by agobL** [reply]
After looking at the example, I succeeded to have my code running with JIT. Thanks a lot !
